### PR TITLE
fix: creating a Parse.File with base64 string fails for some file types

### LINKED ISSUE #1466: Wrong processing of a base64 string in accordance to RFC 2397 when calling new Parse.File
### New Issue Checklist

- [x] I am not disclosing a [vulnerability](https://github.com/parse-community/parse-server/blob/master/SECURITY.md).
- [x] I am not just asking a [question](https://github.com/parse-community/.github/blob/master/SUPPORT.md).
- [x] I have searched through [existing issues](https://github.com/parse-community/Parse-SDK-JS/issues?q=is%3Aissue).
- [x] I can reproduce the issue with the latest versions of [Parse Server](https://github.com/parse-community/parse-server/releases) and the [Parse JS SDK](https://github.com/parse-community/Parse-SDK-JS/releases). 

### Issue Description

It is common practice to specify other parameters in the date URI such as `;filename="parse.pdf"`.
I discovered this issue after generating a pdf in the browser and tried to create a new file using `Parse.File(...)`.

I created the following test to point out the error:
```js
it('can extract data type from base64 with a filename parameter', () => {
  const file = new ParseFile('parse.pdf', {
    base64: 'data:application/pdf;filename=parse.pdf;base64,ParseA==',
  });
  expect(file._source.base64).toBe('ParseA==');
  expect(file._source.type).toBe('application/pdf');
});
```

With the following results:
image

The main problem is that the currently used regex doesn't take into account that there may be other parameters (such as `filename`). It takes into account only `charset`.
```js
const dataUriRegexp = /^data:([a-zA-Z]+\/[-a-zA-Z0-9+.]+)(;charset=[a-zA-Z0-9\-\/]*)?;base64,/;
```

### Steps to reproduce

Create a new Parse.File with data starting with `data:application/pdf;filename=test.pdf;base64,`

### Actual Outcome

`new Parse.File(...)` throws an error at line 150 in ParseFile.js.

### Expected Outcome

`new Parse.File(...)` should return a new object even when there are other parameters specified.

###  Environment

Client
- Parse JS SDK version: `3.4.1`

### Logs

- n/a

### PR DESCRIPTION
**Pull request by RazvanCristian**
### New Pull Request Checklist

- [x] I am not disclosing a [vulnerability](https://github.com/parse-community/parse-server/blob/master/SECURITY.md).
- [x] I am creating this PR in reference to an [issue](https://github.com/parse-community/Parse-SDK-JS/issues?q=is%3Aissue).

### Issue Description

Related issue: [#1466](https://github.com/parse-community/Parse-SDK-JS/issues/1466)

### Approach

Reworked logic for base64 formatting in ParseFile.js.

### TODOs before merging

- [x] Add tests
- [x] Add entry to changelog

### CODE REVIEW

File: changelogs/CHANGELOG_release.md
Comment (mtrezza) [inline changelogs/CHANGELOG_release.md]: Changelog entries are not required anymore, please revert any changes to the file.

File: src/__tests__/ParseFile-test.js
Comment (mtrezza) [inline src/__tests__/ParseFile-test.js]: Does this throw an error without the fix? Can you please revert changes to the existing test and add a new test with a description that reflects what this PR fixes? For example, why does it throw an error without the fix and what is it specifically that we’re testing? The PR title should also reflect that.

**Review by mtrezza — CHANGES_REQUESTED** [review]

File: src/__tests__/ParseFile-test.js
Comment (RazvanCristian) [reply src/__tests__/ParseFile-test.js]: Hi, the description was made in a hurry, I updated the issue to better describe the problem. (see #1466)  
I also updated the tests:
image

File: src/ParseFile.js
Comment (mtrezza) [inline src/ParseFile.js]: This is a less strict string parsing than the previous regex. What are the benefits of removing the original regex and could a less strict parsing create potential vulnerabilities?

File: src/__tests__/ParseFile-test.js
Comment (mtrezza) [inline src/__tests__/ParseFile-test.js:78]: Not sure whether this was deliberately a `txt` extension to demonstrate that the image type is derived from the data, not from the extension?

Comment (RazvanCristian) [reply src/__tests__/ParseFile-test.js:78]: It doesn't matter that much. The type is never taken from the extension. In other tests the extension corresponds to the specified type.

File: src/ParseFile.js
Comment (RazvanCristian) [reply src/ParseFile.js]: I added validation to check if the string follows the scheme of a URI or if it is base64 encoded data.

File: src/ParseFile.js
Comment (rubenatorio) [inline src/ParseFile.js]: According to RFC 2397 if there is no type shouldn't it default to `text/plain`?

```
If <mediatype> is omitted, it
   defaults to text/plain;charset=US-ASCII.  As a shorthand,
   "text/plain" can be omitted but the charset parameter supplied.
```

Comment (RazvanCristian) [reply src/ParseFile.js]: You're right, I've set the default type to `text/plain`.

**Review by mtrezza — CHANGES_REQUESTED** [review]
Tests failing, please request another review from `@parse-community/js-sdk-review` when this is ready.

**Comment by RazvanCristian** [comment]
> Tests failing, please request another review from `@parse-community/js-sdk-review` when this is ready.

Tests are failing because you're using node `14.19.1`, while the current LTS version is node `16.14.2`. 
image
`||=` operator was added in `15.0.0`. [Logical OR assignment (||=)](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Logical_OR_assignment)

**Comment by mtrezza** [comment]
Can you adapt the code to pass? We won't change the node engine version at this point. Node 12 and 14 are still under LTS.

**Comment by RazvanCristian** [comment]
Done.

**Review by mtrezza — CHANGES_REQUESTED** [review]
Tests still fail, could you take a look? And please let me know if the new PR title is correct.

**Comment by RazvanCristian** [comment]
> Tests still fail, could you take a look? And please let me know if the new PR title is correct.

Sorry for taking so long, I had some work to do. 
I updated the PR and checked that it passes the tests with version `v14.19.1`. It passes all the tests and there should be no other problems.

**Comment by mtrezza** [comment]
I changed the PR title because the previous one was difficult to read. It seems you changed it back. What's incorrect or missing for the title:

> fix: creating a Parse.File with base64 string fails for some file types

File: src/ParseFile.js
Comment (mtrezza) [inline src/ParseFile.js]: Could you rephrase the error message (if you think it makes sense) to:

> Cannot create a Parse.File without valid data URIs or base64 encoded data.

It also seems that this new error is missing a test case, could you add one?

**Review by mtrezza — CHANGES_REQUESTED** [review]

**Review by mtrezza — APPROVED** [review]
Looks good to me! Thanks for this fix, let's just wait for the CI to pass.

**Comment by mtrezza** [comment]
@rubenatorio Do you want to do a final review before we merge?

**Comment by parseplatformorg** [comment]
🎉 This change has been released in version [3.4.3-alpha.1](https://github.com/parse-community/Parse-SDK-JS/releases/tag/3.4.3-alpha.1)

**Comment by parseplatformorg** [comment]
🎉 This change has been released in version [3.4.3-beta.1](https://github.com/parse-community/Parse-SDK-JS/releases/tag/3.4.3-beta.1)

**Comment by parseplatformorg** [comment]
🎉 This change has been released in version [3.4.3](https://github.com/parse-community/Parse-SDK-JS/releases/tag/3.4.3)

**Comment by Saquib764** [comment]
It seems there is still issue with `Parse.File` while creating base64 file.

I am generating data URL from canvas-
```
let url = canvas.toDataURL("image/png");

const file = new Parse.File(`shading_texture.png`, { base64: url.split('base64,')[1] });
await file.save()
```

Upon running this script, the frontend freezes for a couple of minutes and then fails. 

The exact same script was working with version `3.4.2`

I also tried the following script to extract base64 string. No success.
```
let base64 = url.replace('data:', '').replace(/^.+,/, '')
```

**Comment by mtrezza** [comment]
Could you open a new issue?

**Comment by samateja** [comment]
this fix is not working when i upload a video and convert the video to base64. The file size is around 17mb, do you suggest any other solution?

**Comment by FransGH** [comment]
See #1532

**Pull request merged**
